### PR TITLE
refactor(cache): upgrade foyer to 0.12 and related deps

### PR DESCRIPTION
**Pull request by MrCroxx**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Changes:

- Upgrade foyer to 0.12
- Upgrade hashbrown to 0.15. (https://github.com/rust-lang/hashbrown/issues/564)
- Fix typos with latest typo-cli.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by hzxa21 — APPROVED** [review]

**Comment by hzxa21** [comment]
compilation fails for hashbrown 0.15: https://buildkite.com/risingwavelabs/pull-request/builds/59455#01927058-8ce4-4e74-8ed1-f5c2f047df6e

File: src/storage/src/store_impl.rs [hunk 1]
```diff
@@ -643,27 +645,32 @@ impl StateStoreImpl {
...
+                    .with_runtime_options(opts.meta_file_cache_runtime_config.clone())
+                    .with_large_object_disk_cache_options(
+                        LargeEngineOptions::new()
+                            .with_indexer_shards(opts.meta_file_cache_indexer_shards)
+                            .with_flushers(opts.meta_file_cache_flushers)
+                            .with_reclaimers(opts.meta_file_cache_reclaimers)
+                            .with_buffer_pool_size(
+                                opts.meta_file_cache_flush_buffer_threshold_mb * MB,
+                            ) // 128 MiB
+                            .with_clean_region_threshold(
```
Comment (Li0k) [inline src/storage/src/store_impl.rs:668]: just a question: Is the value of `clean_region_threshold` 1.5 times the value of `reclaimers`? Is this a specially tuned parameter for RW environments?

**Review by Li0k — APPROVED** [review]
LGTM

File: src/storage/src/store_impl.rs (hunk 1 above)
Comment (MrCroxx) [reply src/storage/src/store_impl.rs:668]: Yes, it is x1.5. But I think it may not affect that much.

**Pull request merged**